File: core/src/com/qtech/bubbleblaster/init/Abilities.java
package com.qtech.bubbleblaster.init;

import com.qtech.bubbleblaster.QInternalAddon;
import com.qtech.bubbleblaster.ability.TeleportAbility;
import com.qtech.bubbleblaster.common.ability.AbilityType;
import com.qtech.bubbleblaster.registry.DeferredRegister;
import com.qtech.bubbleblaster.registry.Registers;
import com.qtech.bubbleblaster.registry.object.RegistryObject;

import java.util.function.Supplier;

@SuppressWarnings({"rawtypes", "unused", "SameParameterValue"})
public class Abilities {
    public static final DeferredRegister<AbilityType> ABILITY_TYPES = DeferredRegister.create(QInternalAddon.ADDON_ID, Registers.ABILITIES);
    public static final RegistryObject<AbilityType<TeleportAbility>> TELEPORT_ABILITY = register("teleport", () -> new AbilityType<>(TeleportAbility::new));

    private static <T extends AbilityType<?>> RegistryObject<T> register(String name, Supplier<T> supplier) {
        return ABILITY_TYPES.register(name, supplier);
    }

}

File: src/main/kotlin/qtech/bubbles/init/Abilities.kt
package qtech.bubbles.init

import qtech.bubbles.BBInternalMod
import qtech.bubbles.ability.TeleportAbility
import qtech.bubbles.common.ability.AbilityType
import qtech.bubbles.registry.DeferredRegister
import qtech.bubbles.registry.Registers
import qtech.bubbles.registry.`object`.RegistryObject
import java.util.function.Supplier

object Abilities {
    val ABILITY_TYPES: DeferredRegister<AbilityType<*>> = DeferredRegister.create(BBInternalMod.ADDON_ID, Registers.ABILITIES)
    val TELEPORT_ABILITY: RegistryObject<AbilityType<TeleportAbility>> = register("teleport") { AbilityType { TeleportAbility() } }
    private fun <T : AbilityType<*>> register(name: String, supplier: Supplier<T>): RegistryObject<T> {
        return ABILITY_TYPES.register(name, supplier)
    }
}